### ISSUE TITLE
Unexpected difference

### ISSUE THREAD

**Issue by FHantke** [rugleb/JsonCompare#29]
Hi,
thank you for the great package. I use it to compare HTML DOM trees, however, in some situations the package outputs wrong diffs. Attached you can find two [JSON files](https://github.com/rugleb/JsonCompare/files/9171965/examples.zip) that are smaller parts of a bigger DOM tree. At the same position, the first JSON contains the value `THIS IS THE CORRECT VALUE`, the second one `THIS IS THE WRONG VALUE`.
```
r1["childNodes"][0]["childNodes"][0]["childNodes"][0]["childNodes"][0]["childNodes"][0]["childNodes"][0]["nodeValue"]
THIS IS THE CORRECT VALUE
```
Both JSONs also contain the value `THIS IS NOT DIFFERENT` at the same position:
```
r1["childNodes"][1]["childNodes"][0]["childNodes"][0]["childNodes"][1]["childNodes"][0]["childNodes"][0]["nodeValue"]
THIS IS NOT DIFFERENT
```

When comparing both objects, I expect the outcome to tell me that the wrong and correct value are different. However, the result is the following, stating that the correct value is different from the "not different" value.

```
{'childNodes': {'_content': {0: {'childNodes': {'_content': {0: {'childNodes': {'_length': {'_message': 'Lengths not equal. Expected <1>, received: <2>',
         '_expected': 1,
         '_received': 2},
        '_content': {0: {'childNodes': {'_length': {'_message': 'Lengths not equal. Expected <1>, received: <3>',
            '_expected': 1,
            '_received': 3},
           '_content': {0: {'childNodes': {'_length': {'_message': 'Lengths not equal. Expected <1>, received: <2>',
               '_expected': 1,
               '_received': 2},
              '_content': {0: {'attributes': {'href': {'_message': 'Values not equal. Expected: </whats-new/latest-activity>, received: </search/?type=profile_post>',
                  '_expected': '/whats-new/latest-activity',
                  '_received': '/search/?type=profile_post'},
                 'rel': {'_message': 'Key does not exists. Expected: <rel>',
                  '_expected': 'rel',
                  '_received': None},
                 'data-nav-id': {'_message': 'Values not equal. Expected: <latestActivity>, received: <searchProfilePosts>',
                  '_expected': 'latestActivity',
                  '_received': 'searchProfilePosts'}},
                'childNodes': {'_content': {0: {'nodeValue': {'_message': 'Values not equal. Expected: <THIS IS THE CORRECT VALUE>, received: <THIS IS NOT DIFFERENT>',
                    '_expected': 'THIS IS THE CORRECT VALUE',
                    '_received': 'THIS IS NOT DIFFERENT'}}}}}}}}}}}}}}}}}}}}
```

I had not the time yet to look deeper into the issue, but it seems that some indexing in the package code is wrong. Somehow, the code compares `r1["childNodes"][0]` with `r2["childNodes"][1]`. If I remove `["childNodes"][1]` from both JSONs, the outcome is correct. Furthermore, if I manually compare `r1["childNodes"][0]` and `r2["childNodes"][1]`, the outcome is the same as the one printed above.
```
r10 = r1["childNodes"][0]
r21 = r2["childNodes"][1]
Compare().check(r10, r21)
```

I hope that my examples help to find the issue.
Thank you for your help!

**Comment by FHantke** [comment]
A smaller example:
```python
from jsoncomparison import Compare
j1 = {
  'children': [
    {'a': [{'b': 'xxxx'}]},
    {'a': [{'b': 'yyyy'}]}
  ]
}

j2 = {
  'children': [
    {'a': [{'b': 'zzzz'}]},
    {'a': [{'b': 'yyyy'}]}
  ]
}
Compare().check(j1, j2)
```

Outcome:
```python
{'children': {'_content': {0: {'a': {'_content': {0: {'b': {'_message': 'Values not equal. Expected: <xxxx>, received: <yyyy>',
        '_expected': 'xxxx',
        '_received': 'yyyy'}}}}}}}}
```

**Comment by FHantke** [comment]
I looked deeper into it and tried an even smaller example:
```
expected = [
    {'a': "xxx"},
    {'a': "yyy"}
  ]
actual = [
    {'a': "zzz"},
    {'a': "yyy"}
  ]
Compare().check(expected, actual)
```
Outcome:
```
{'_content': {0: {'a': {'_message': 'Values not equal. Expected: <xxx>, received: <yyy>',
    '_expected': 'xxx',
    '_received': 'yyy'}}}}
```

The problem is the following code:
https://github.com/rugleb/JsonCompare/blob/2d82fc21f532b9faf8d1d3d1da00c9894249592d/jsoncomparison/compare.py#L156-L164

In the final comparison, the values are the following:
```
dd = {'a': {'_message': 'Values not equal. Expected: <xxx>, received: <yyy>', '_expected': 'xxx', '_received': 'yyy'}}
d  = {'a': {'_message': 'Values not equal. Expected: <xxx>, received: <zzz>', '_expected': 'xxx', '_received': 'zzz'}}
```

Since the length of both is 1, the algorithm decides to take `dd` while the correct result would be `d`. Unfortunately, I'm not good with algorithmic and don't know how to fix this issue.

**Comment by rugleb** [comment]
Thanks for u report. I will think how to fix it later.
I have plans to completely rewrite this project, since it still consists of shit code.

**Comment by MiyamotoAkira** [comment]
I will have a look into this one this weekend.

**Comment by rugleb** [comment]
@FHantke Hi!
Can u see #37? Does it fix the problem?

**Comment by rugleb** [comment]
Closed in #37

**Comment by rugleb** [comment]
Yes, I will do it one of these days
